Create PresStudyTabulations.sas

RAGarMir opened the pull request:
This is the progress I have had so far on creating the tabulations for the first 3 tables. I have the shell of the code done and I am working through the errors. Currently working on errors when merging datasets Project_AOB & Realprop.Cama_parcel_2023_05 by SSL.

ptatian commented on Prog/PresStudyTabulations.sas:
```diff
@@ -0,0 +1,369 @@
+  by SSL;
+ 	 if inProject and inCama;
+   
+run;
```
`ERROR: BY variables are not properly sorted on data set WORK.PROJECT_AOB.`

When doing a MERGE, all data sets must be sorted by the BY variables. Use [Proc Sort](https://documentation.sas.com/doc/en/proc/1.0/p0ha9ymifyaqldn14m2xlqw252wa.htm) to sort the Project_AOB data set by SSL before the data step that does the merge.

ptatian commented on Prog/PresStudyTabulations.sas:
```diff
@@ -0,0 +1,369 @@
+run;
+
+proc summary data=Project_Age_Of_Building 
+  by nlihc_id;
```
`ERROR 22-322: Syntax error, expecting one of the following: ;, (, ...`

Missing a semicolon at the end of the Proc Summary statement.

ptatian commented on Prog/PresStudyTabulations.sas:
```diff
@@ -0,0 +1,369 @@
+    sum='Assisted Units' * ( all='\b Total' ProgCat=' ' ) * mid_asst_units=' '
+    ;
+  format ProgCat ProgCat.;
+  format value year_built (notsorted)
```
`ERROR 22-322: Syntax error, expecting one of the following: a name, a format name, ...`

The code to create a new user format needs to be a Proc Format (not a Format statement) which is separate from and comes before the Proc Tabulate code. Please refer back to the code I included in the issue. You then assign the year_built format you created to the AYB variable using a Format statement as part of the Proc Tabulate, such as is being done with `format ProgCat ProgCat.;`.

[You can look here](https://www.statology.org/sas-proc-format/) for other examples of using Proc Format.

ptatian reviewed (commented):
@RAGarMir Please see my comments on the errors you are getting. Let me know if you are still having challenges.

ptatian commented:
@RAGarMir Regarding this error.
```
399        ***Getting the Message that this statement is out of order***;
400        proc summary data=Project_Age_Of_Building;
401          by nlihc_id;
402          output out=Project_Building_Age;
403            min(AYB)=;
               ___
The SAS System

               180
ERROR 180-322: Statement is not valid or it is used out of proper order.
404        run;
```
The problem is that `min(AYB)=` is part of the output statement. Therefore, you should not have a semicolon after `Project_Building_Age`.

RAGarMir commented:
Hello @ptatian I removed that semicolon. I still kept getting an error at the Proc Summary statement saying the data set was not sorted properly. I changed "by nlihc_id" to "by SSL" and that seemed to get rid of the error, but, unless I was reading your instructions wrong, those said to sort by NLIHC_ID at this stage. Is there something else I may be missing here?

ptatian commented:
Hi @RAGarMir. Yes, sorry, I overlooked that. The input data set does need to be sorted by the BY variables. But we want to summarize by project (nlihc_id) and not by parcel, so changing to BY SSL is not what we want. 

There are two possible solutions. One is to use [Proc Sort](https://documentation.sas.com/doc/en/pgmsascdc/v_059/proc/n0sv3pcvpllvh9n1521qlriu6u4n.htm) to sort the data set Project_Age_Of_Building before the Proc Summary step. 

A second option is to change Proc Summary step as follows:
```
proc summary data=Project_Age_Of_Building nway;
  class nlihc_id;
```
Changing BY to CLASS removes the need to have the input data sorted. But CLASS by itself produces multiple summary levels. Adding the NWAY option to Proc Summary restricts to just the most basic summary. 

If the only reason we need Project_Age_Of_Building sorted is for the Proc Summary, then I would probably go with the second option.

ptatian commented:
Just to clarify, the rest of the Proc Summary step would be the same as before. Only the two lines I highlighted above need to be changed.

RAGarMir commented:
@ptatian I created and ran the code to create the table by owner type. The code tan fine without any errors, but the desired table is not part of the output, I am not sure how to debug in this situation.

ptatian commented:
@RAGarMir The proc tabulate needs to be placed before the ods rtf close statement. Only output from code between the first ods rtf… statement and the ods rtf close statement will be directed to the output file.

RAGarMir commented:
Hello @ptatian I currently have the ods rtf close statement currently comes after the proc tabulate statement but I am still not getting the output of the table in the word document.

ptatian commented:
Hi @RAGarMir Yes. You do have the Proc Tabulate in the right place now, but the problem is the Proc Tabulate isn't producing any output. If you look at the LOG, you will see these messages after the last Proc Tabulate.

```
NOTE: No observations were selected from data set WORK.PROJECT_ASSISTED_UNITS.
NOTE: There were 0 observations read from the data set WORK.PROJECT_ASSISTED_UNITS.
      WHERE (ProgCat not = .) and (not MISSING(parcel_owner_type));
```

Since no data were selected, no table was produced. 

So, the next step would be to go back to where PROJECT_ASSISTED_UNITS was created and see what might be going on. 

If we add parcel_owner_type to the list of freqvars= in the %File_info() macro, we get this in the output. 

```
parcel_
owner_                              Cumulative    Cumulative
type       Frequency     Percent     Frequency      Percent
------------------------------------------------------------
                541      100.00           541       100.00  
```

This tells us that all the values of parcel_owner_type are blank. Blank equates to "missing" for character variables in SAS and, therefore, the condition `not( missing( parcel_owner_type ) )` in the Proc Tabulate where statement is never true. 

So, the next step is to figure out what is going wrong with parcel_owner_type. If we look in the LOG after the Project_Owner_Summary data step, we see messages like this:

```
NOTE: Variable parcel_owner_type_111 is uninitialized.
NOTE: Variable parcel_owner_type_045 is uninitialized.
NOTE: Variable parcel_owner_type_115 is uninitialized.
etc.
```

This means that the variables parcel_owner_type_111, parcel_owner_type_045, etc. were never set to any value before they were used. We need to continue to trace back to see where the problem emerged. The step before this one was the Proc Summary, so let's look at that. 

```
***Summarizing new variables by Nlihc_id***;
proc summary data=parcel_owner_type nway; 
  class nlihc_id;
  output out=parcel_owner_type;
run;
```

As noted in the comment, the purpose of Proc Summary was to summarize the new variables, but you didn't include the new variables in the Proc Summary. You need to provide a VAR statement to do this. Also, you didn't provide any statistic keyword with the output statement, which means SAS will use SUM= as the default. But, we actually want MAX= here so that the values of the summarized variables will be equal to either 1 or missing. 

```
proc summary data=parcel_owner_type nway; 
  class nlihc_id;
  var parcel_owner_type_111 Variable parcel_owner_type_045 parcel_owner_type_115 etc.;
  output out=parcel_owner_type max=;
run;
```

Once that is fixed, the values of parcel_owner_type should be assigned properly. You should **check the data**, however, to make sure that the code is working the way we want. That is, you should look not only at the final output, but at some of the intermediate data sets so that you can be sure that things are working properly. You can find information on this in the [Examining Intermediate Results section](https://github.com/NeighborhoodInfoDC/SAS/blob/master/Guide/Debugging%20programs.mkdn#Intermediate) of the debugging guide.

ptatian commented:
I went through a lengthy description above so that you could see my process for tracking down errors. You need to **work your way backward** through the code to identify where the problem is occuring. Look at **intermediate results** (data sets, output) to see that everything is working as expected. Try to trace the problem back to its earliest point. That is the first part of finding a solution.

I linked to one part of it above but you should take a look again at the [SAS Debugging Guide](https://github.com/NeighborhoodInfoDC/SAS/blob/master/Guide/Debugging%20programs.mkdn). And [this tutorial](https://libguides.library.kent.edu/SAS/ViewData) explains how to view data sets either in the SAS Windows Viewtable window with Proc Print.

RAGarMir commented:
@ptatian I made the requested changes, additions, and fixes.

RAGarMir commented:
@ptatian the latest commit should have the code I added for the tables on neighborhood characteristics, earliest expiration dates, and size of development

RAGarMir commented:
@ptatian I consolidated projects and assisted units into one table for Sec8 and Lihtc tables

ptatian commented:
Hi @RAGarMir. Thank you for all your work on this. I made a few adjustments to the expiration tables, and also added code for outputting lists of expiring Sec 8 and LIHTC projects. In making these changes, I noticed that the year built/renovated tables were not using the project-level data, but a data set that had multiple records for each project. I fixed that by merging the summary AYB and EYB variables back onto the Project_assisted_units data. 

I'll approve all the changes for the review. I'd like you to finish this up, please, by merging the branch with main in the pull request, running the Sync.ps1 file in Windows Powershell to update the code on SAS1, and then doing one last batch submit of the program on SAS1. When finished, please copy the full pathname of the LOG file on SAS1 file to issue #473 and close the issue. 

Let me know if you have any questions. Thank you!

ptatian reviewed (approved):
Changes approved. Ready to be finalized.
